### DISCUSSION TITLE
Adding roles to user

### DISCUSSION THREAD

**Discussion by geric** [nextauthjs/next-auth#2929] [Help] [answered]
**Question:**
Hi, I really love the simplicity and abstraction of next-auth.

I am trying to implement a role-based access to my application. 
In order to achieve this, I wanted to add role to my user.

In next-auth **version 3.29.0** I was able to do it using TypeORM adapter:

```javascript
adapter: Adapters.TypeORM.Adapter(process.env.DATABASE_URL, {
      models: {
        User: models.User, // custom user model
      },
    }),
```

In the latest version, using **4.0.0-beta.4**, I wanted to try and use MongoDBAdapter instead of TypeORMLegacyAdapter.
I am unable to find a solution to modify the default user schema using the MongoDBAdapter.

Right now, I was able to add role by using the event createUser
```javascript
events: {
      createUser: async (message) => {
        const client = await clientPromise;
        const usersCollection = client.db().collection('users');
        await usersCollection.updateOne(
          { email: message.user.email },
          { $set: { role: 'USER' } }
        );
      },
    },
```
This solution works but I want to modify the default User scheme to set a default role and avoid DB update upon user creation.

**Comment by balazsorban44** [comment, marked as answer]
With v4, the only thing you have to modify is what your `profile` callback returns, no need to maddle with the adapter code or events!

See: https://next-auth.js.org/adapters/models#user

So something like this is enough:
```js
profile(profile) {
  return {
    role: "USER",
    id: profile.id,
    ... // etc.
  }
}
```

**Reply by geric** [reply]
Awesome! I missed this part in the v4 documentation. 
I've added the profile callback in my OAuth Provider and it's works. 
Additional question: I can see that the profile callback is only available on OAuth providers. What if I wanted to use only email provider, what is the best way to add the custom user details?

**Reply by eakl** [reply]
Same question. How can we modify the user data for the email provider? It seems to be hard coded in the callback handler

**Reply by raulmarindev** [reply]
hi @eakl, were you able to solve it?

**Reply by mabdullahadeel** [reply]
> With v4, the only thing you have to modify is what your `profile` callback returns, no need to maddle with the adapter code or events!
> 
> See: https://next-auth.js.org/adapters/models#user
> 
> So something like this is enough:
> 
> ```js
> profile(profile) {
>   return {
>     role: "USER",
>     id: profile.id,
>     ... // etc.
>   }
> }
> ```

@balazsorban44 awesome work with the library. But I wanted to know if it would be type-safe? I mean if I modify and add a new field which in this case is `role`, then `callback` functions e.g `signIn` expecting profile as parameter will not be able to infer the correct types.
I have searched for ways to do it with Prisma but couldn't find the solution. 

If there is some way around it except defining types manually, then please mention.

Thanks & Keep up the good work 👐

**Reply by joepetrillo** [reply]
@geric @eakl @raulmarindev were any of you able to figure this out using email login?
